File: easy/RearrangeCharToMakeStr.js
// You are given two 0-indexed strings s and target. You can take some letters from s and rearrange them to form new strings.

// Return the maximum number of copies of target that can be formed by taking letters from s and rearranging them.

const strToObj = function (s) {
  const letters = {};
  for (let i = 0; i < s.length; i++) {
    letters[s[i]] = letters[s[i]] + 1 || 1;
  }
  return letters;
}
const rearrangeCharacters = function (s, target) {
  const sLet = strToObj(s);
  const tLet = strToObj(target);
  let copies = Infinity;
  for (let char in tLet) {
    if (sLet[char]) {
      const num = Math.floor(sLet[char] / tLet[char]);
      if (copies > num) {
        copies = num;
      }
    } else {
      return 0;
    }
  }
  return copies;
};

console.log(rearrangeCharacters('ilovecodingonleetcode', 'code') === 2)

File: easy/RearrangeCharToMakeStr.py
# You are given two 0-indexed strings s and target. You can take some letters from s and rearrange them to form new strings.

# Return the maximum number of copies of target that can be formed by taking letters from s and rearranging them.

import math


def strToObj(s):
    obj = {}
    for char in s:
        obj[char] = obj.get(char, 0) + 1
    return obj


def rearrangeCharacters(s, target):
    inputObj = strToObj(s)
    targetObj = strToObj(target)
    copies = math.inf
    for char in targetObj:
        if char in inputObj:
            copies = min(copies, inputObj[char] // targetObj[char])
        else:
            return 0
    return copies


print(rearrangeCharacters('ilovecodingonleetcode', 'code') == 2)
